Bug 1177519 - Autoclassify feature

camd opened the pull request:
<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/mozilla/treeherder/1234)
<!-- Reviewable:end -->

edmorley commented:
Following on from our Vidyo call yesterday, could you:
* Drop the flake8 exclusion changes in setup.cfg (since the issue it's finding is fixed in Django 1.9 and excluding the migrations means we lose coverage for manually created/tweaked migrations)
* Look into whether the eslint ES6 change makes sense for us (ie whether our tooling supports it)
* Break out the Procfile changes (think after all it might be worth it)
* Break out the sample.local.conf.js change
* s/TREEHERDER_AUTOCLASSIFY_JOBS/ENABLE_AUTOCLASSIFICATION/ perhaps? (We're moving away from the TREEHERDER prefix)
* Update the added django-rest-framework-filters to 0.6.0
* Add any of the commit messages from the unsquashed branch as inline code comments if they are something that's not otherwise clear

Nice job on all this - it's really starting to come together :-)

maurodoglio commented:
Reviewed 1 of 1 files at r1, 2 of 2 files at r2, 22 of 22 files at r3.
Review status: 24 of 47 files reviewed at latest revision, 21 unresolved discussions.

---

*[treeherder/model/derived/jobs.py, line 432 \[r3\]](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-K7l9ITJVfY5MvRDlluI:-K7l9ITJVfY5MvRDlluJ:-2086348510) ([raw file](https://github.com/mozilla/treeherder/blob/308042f800d2d5f34865f123be04b5659457efb5/treeherder/model/derived/jobs.py#L432)):*
Keeping in mind that we want to stop using JobsModel and datasource one day, it would be good to move this code on the reference data side of things. If you convert the job_id to a job_guid early, you don't really need to do much work with JobsModel. Also, I guess the info that you are updating here is mostly part of the reference schema. Anyway, I'll leave the final word on this to you and @edmorley as you have a better view of the whole picture.

---

*[treeherder/model/derived/jobs.py, line 590 \[r3\]](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-K7l8wR-9pRH97uDnydi:-K7l8wR0YdFFHA5fa3co:1235685911) ([raw file](https://github.com/mozilla/treeherder/blob/308042f800d2d5f34865f123be04b5659457efb5/treeherder/model/derived/jobs.py#L590)):*
I see this method as a better fit for ArtifactsModel

---

*[treeherder/model/migrations/0008__add_best_fields_to_failureline.py, line 15 \[r2\]](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-K7kzvSrNgnUxhILjl1v:-K7kzvSs1xQMkPQeMlYx:1553605144) ([raw file](https://github.com/mozilla/treeherder/blob/663ca39a8b491010396479022d0793a67b26798e/treeherder/model/migrations/0008__add_best_fields_to_failureline.py#L15)):*
I don't know if we have some classification data that we want to keep, but this migration will likely drop any existing info on which classification is the best one. Maybe we should have 2 different migrations, one for the addition of best_classification (together with eventual data migrations) and one for the removal of is_best?

---

*[treeherder/model/models.py, line 617 \[r2\]](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-K7kn6C3_PnIpKF4sWGM:-K7kn6C3_PnIpKF4sWGN:1928144179) ([raw file](https://github.com/mozilla/treeherder/blob/663ca39a8b491010396479022d0793a67b26798e/treeherder/model/models.py#L617)):*
This function name suggests that there's a verification process going on inside it. What is actually doing is marking the best classification as verified. Can you please change the name to better describe its behavior?

---

*[treeherder/model/models.py, line 630 \[r2\]](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-K7krdNAkhnl_8IvEsE2:-K7krdNAkhnl_8IvEsE3:-1184605988) ([raw file](https://github.com/mozilla/treeherder/blob/663ca39a8b491010396479022d0793a67b26798e/treeherder/model/models.py#L630)):*
Can you please add some docstrings to this function?
After looking at it for 10 minutes I know what it does, but it was not obvious.
This is matching a failure line with bug suggestions, right?

---

*[treeherder/model/models.py, line 665 \[r2\]](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-K7ky-o22yF6-IJ3pzpp:-K7ky-o22yF6-IJ3pzpq:210014173) ([raw file](https://github.com/mozilla/treeherder/blob/663ca39a8b491010396479022d0793a67b26798e/treeherder/model/models.py#L665)):*
This function seems to do what [the first() queryset method](https://docs.djangoproject.com/en/1.9/ref/models/querysets/#first) does.
You should either remove it or replace its body with
```python
return Bugscache.objects.first(id=self.bug_number)
```

---

*[treeherder/model/models.py, line 677 \[r2\]](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-K7kzMfBxK_P1exyK6aE:-K7kzMfBxK_P1exyK6aF:258180615) ([raw file](https://github.com/mozilla/treeherder/blob/663ca39a8b491010396479022d0793a67b26798e/treeherder/model/models.py#L677)):*
I feel like I already reviewed this code from jgraham, am I wrong?

---

*[treeherder/model/models.py, line 728 \[r2\]](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-K7klS7WrcRmaQJZvC-s:-K7klS7XALEut1Uwl7mh:-571277238) ([raw file](https://github.com/mozilla/treeherder/blob/663ca39a8b491010396479022d0793a67b26798e/treeherder/model/models.py#L728)):*
I'm not sure you need @classmethod anymore since you are not using cls anymore

---

*[treeherder/model/models.py, line 731 \[r2\]](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-K7kl0i2FBaQsTrB98i0:-K7kl0i3j1r-RQ4Tf5Fm:256623964) ([raw file](https://github.com/mozilla/treeherder/blob/663ca39a8b491010396479022d0793a67b26798e/treeherder/model/models.py#L731)):*
Shouldn't this return
```python
return dest[matcher_cls.__name__]
```
instead?

---

*[treeherder/webapp/api/classifiedfailure.py, line 28 \[r3\]](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-K7lCCUeRfqrNpDT3f0a:-K7lCCUf2nfObI5N7YJT:2146776171) ([raw file](https://github.com/mozilla/treeherder/blob/308042f800d2d5f34865f123be04b5659457efb5/treeherder/webapp/api/classifiedfailure.py#L28)):*
Why did you implement this method instead of using the base one?

---

*[treeherder/webapp/api/classifiedfailure.py, line 32 \[r3\]](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-K7lCJpukAhby-E1ykbT:-K7lCJpukAhby-E1ykbU:1492812845) ([raw file](https://github.com/mozilla/treeherder/blob/308042f800d2d5f34865f123be04b5659457efb5/treeherder/webapp/api/classifiedfailure.py#L32)):*
In cases like this we should return a [NotFound exception](http://www.django-rest-framework.org/api-guide/exceptions/#notfound) which translates to a 404 error.

---

*[treeherder/webapp/api/classifiedfailure.py, line 36 \[r3\]](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-K7lGnHKXfeZG8aiMMzE:-K7lGnHLkOPHxmY-tr3m:1998107676) ([raw file](https://github.com/mozilla/treeherder/blob/308042f800d2d5f34865f123be04b5659457efb5/treeherder/webapp/api/classifiedfailure.py#L36)):*
This creation logic should be part of a serializer. We already have a ClassifiedFailureSerializer and you can implement the save method on that. You will also need a ClassifiedFailureListSerializer to handle the many=True case. Here is the docs on [ListSerializers](http://www.django-rest-framework.org/api-guide/serializers/#listserializer).

---

*[treeherder/webapp/api/classifiedfailure.py, line 45 \[r3\]](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-K7lHxsrAuQSEkJn3h3u:-K7lHxsrAuQSEkJn3h3v:-1799423446) ([raw file](https://github.com/mozilla/treeherder/blob/308042f800d2d5f34865f123be04b5659457efb5/treeherder/webapp/api/classifiedfailure.py#L45)):*
Either use ClassifiedFailure.objects.create or indent obj.save() under the `else` body please. It will save an insert query when you have a bug number :smiley: .

---

*[treeherder/webapp/api/classifiedfailure.py, line 54 \[r3\]](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-K7lJGAJsiuvAM6KvphV:-K7lJGAJsiuvAM6KvphW:1695174274) ([raw file](https://github.com/mozilla/treeherder/blob/308042f800d2d5f34865f123be04b5659457efb5/treeherder/webapp/api/classifiedfailure.py#L54)):*
The logic here should be something like this https://github.com/miki725/django-rest-framework-bulk/blob/master/rest_framework_bulk/drf3/mixins.py#L26-L35

---

*[treeherder/webapp/api/classifiedfailure.py, line 59 \[r3\]](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-K7lOM6rz2Uet8_2M8Vq:-K7lOM6spZUPnboZANx6:40255228) ([raw file](https://github.com/mozilla/treeherder/blob/308042f800d2d5f34865f123be04b5659457efb5/treeherder/webapp/api/classifiedfailure.py#L59)):*
Please move this code to either a serializer or a model method. 

---

*[treeherder/webapp/api/classifiedfailure.py, line 134 \[r3\]](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-K7lMl6pBafFFfVzEBIF:-K7lMl6pBafFFfVzEBIG:-1258849032) ([raw file](https://github.com/mozilla/treeherder/blob/308042f800d2d5f34865f123be04b5659457efb5/treeherder/webapp/api/classifiedfailure.py#L134)):*
To avoid reinventing wheel we can use [django-rest-framework-bulk](http://www.django-rest-framework.org/api-guide/generic-views/#django-rest-framework-bulk)

---

*[treeherder/webapp/api/classifiedfailure.py, line 146 \[r3\]](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-K7lNT3FSdlO9Q5G_39p:-K7lNT3GPGCE8d1lpu1u:-522124174) ([raw file](https://github.com/mozilla/treeherder/blob/308042f800d2d5f34865f123be04b5659457efb5/treeherder/webapp/api/classifiedfailure.py#L146)):*
maybe we can use a paginator object instead?

---

*[treeherder/webapp/api/failureline.py, line 20 \[r3\]](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-K7lNtfskrwo6l_Ux3NZ:-K7lNtfskrwo6l_Ux3N_:656303176) ([raw file](https://github.com/mozilla/treeherder/blob/308042f800d2d5f34865f123be04b5659457efb5/treeherder/webapp/api/failureline.py#L20)):*
There is no custom logic in here, let's use the base implementation

---

*[treeherder/webapp/api/failureline.py, line 33 \[r3\]](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-K7lO9q14ieMCHQdfBSG:-K7lO9q14ieMCHQdfBSH:40255228) ([raw file](https://github.com/mozilla/treeherder/blob/308042f800d2d5f34865f123be04b5659457efb5/treeherder/webapp/api/failureline.py#L33)):*
Please move this code to either a serializer or a model method. 

---

*[treeherder/webapp/api/failureline.py, line 117 \[r3\]](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-K7lOaSW5JfG9DQg7_NQ:-K7lOaSW5JfG9DQg7_NR:1782652987) ([raw file](https://github.com/mozilla/treeherder/blob/308042f800d2d5f34865f123be04b5659457efb5/treeherder/webapp/api/failureline.py#L117)):*
Let's use http://www.django-rest-framework.org/api-guide/generic-views/#django-rest-framework-bulk for this

---

*[treeherder/webapp/api/serializers.py, line 21 \[r3\]](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-K7lPI6QLzuCRYvqg3p3:-K7lPI6QLzuCRYvqg3p4:1092878374) ([raw file](https://github.com/mozilla/treeherder/blob/308042f800d2d5f34865f123be04b5659457efb5/treeherder/webapp/api/serializers.py#L21)):*
This serializer has the same implementation as the one above, doesn't it? I don't mind if you want to rename the NoOpSerializer to JSONSerializerField

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/1234)*
<!-- Sent from Reviewable.io -->

camd commented:
Review status: 24 of 47 files reviewed at latest revision, 21 unresolved discussions.

---

*[treeherder/model/derived/jobs.py, line 590 \[r3\]](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-K7l8wR-9pRH97uDnydi:-K81jY9-j9VJ0Nr1clDR:-1920539653) ([raw file](https://github.com/mozilla/treeherder/blob/308042f800d2d5f34865f123be04b5659457efb5/treeherder/model/derived/jobs.py#L590)):*
Good call.  Moved over.

---

*[treeherder/model/migrations/0008__add_best_fields_to_failureline.py, line 15 \[r2\]](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-K7kzvSrNgnUxhILjl1v:-K8133s0YZzT542Eebnk:1257102064) ([raw file](https://github.com/mozilla/treeherder/blob/663ca39a8b491010396479022d0793a67b26798e/treeherder/model/migrations/0008__add_best_fields_to_failureline.py#L15)):*
I don't believe the current classification info is important to keep, but James may want to chime in on this, if it's not merged before he gets back.  I sent him an email.

---

*[treeherder/model/models.py, line 617 \[r2\]](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-K7kn6C3_PnIpKF4sWGM:-K81jg9ZevxzbV2UIZ-c:1570848974) ([raw file](https://github.com/mozilla/treeherder/blob/663ca39a8b491010396479022d0793a67b26798e/treeherder/model/models.py#L617)):*
Another good point.  Fixed.  Called it ``set_best_classification_verified``

---

*[treeherder/model/models.py, line 630 \[r2\]](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-K7krdNAkhnl_8IvEsE2:-K81sF8XKShaefDkSi4-:-1760370727) ([raw file](https://github.com/mozilla/treeherder/blob/663ca39a8b491010396479022d0793a67b26798e/treeherder/model/models.py#L630)):*
Yeah, that's correct.  It looks at the Bug Suggestions artifact for searches that match parts of the failure line.  I added a docstring.

---

*[treeherder/model/models.py, line 665 \[r2\]](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-K7ky-o22yF6-IJ3pzpp:-K81tghz3dWwku8dA-Uu:255540908) ([raw file](https://github.com/mozilla/treeherder/blob/663ca39a8b491010396479022d0793a67b26798e/treeherder/model/models.py#L665)):*
I don't see anywhere that uses the ``bug()`` function of this class, but it is probably there for a reason.  James knows better than I.  But I replaced the ``get`` with ``first`` as you suggested. 

---

*[treeherder/model/models.py, line 677 \[r2\]](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-K7kzMfBxK_P1exyK6aE:-K81u-GoZXmQQkdrpzVr:286197855) ([raw file](https://github.com/mozilla/treeherder/blob/663ca39a8b491010396479022d0793a67b26798e/treeherder/model/models.py#L677)):*
Yeah, I think you did review an earlier incarnation.  Sorry for the repeat.  But thanks for the great feedback!

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/1234)*
<!-- Sent from Reviewable.io -->

camd commented:
Review status: 24 of 47 files reviewed at latest revision, 21 unresolved discussions.

---

*[treeherder/model/models.py, line 665 \[r2\]](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-K7ky-o22yF6-IJ3pzpp:-K85kBdcuP0J8UynuOum:564805271) ([raw file](https://github.com/mozilla/treeherder/blob/663ca39a8b491010396479022d0793a67b26798e/treeherder/model/models.py#L665)):*
to be clear, ``first()`` doesn't take params, so I used ``Bugscache.objects.filter(id=self.bug_number).first()``

---

*[treeherder/model/models.py, line 728 \[r2\]](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-K7klS7WrcRmaQJZvC-s:-K85h4HXWPdGCJmOHYdw:-1241107848) ([raw file](https://github.com/mozilla/treeherder/blob/663ca39a8b491010396479022d0793a67b26798e/treeherder/model/models.py#L728)):*
True.  Changed it to @staticmethod

---

*[treeherder/model/models.py, line 731 \[r2\]](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-K7kl0i2FBaQsTrB98i0:-K85h5HIVzhbA7Y4TfLY:-1602310516) ([raw file](https://github.com/mozilla/treeherder/blob/663ca39a8b491010396479022d0793a67b26798e/treeherder/model/models.py#L731)):*
Right you are!  Good catch!  Fixed.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/1234)*
<!-- Sent from Reviewable.io -->

camd commented:
Review status: 18 of 48 files reviewed at latest revision, 21 unresolved discussions.

---

*[treeherder/model/derived/jobs.py, line 432 \[r3\]](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-K7l9ITJVfY5MvRDlluI:-K8RD26wi3b5XzoKIOgr:1613991342) ([raw file](https://github.com/mozilla/treeherder/blob/308042f800d2d5f34865f123be04b5659457efb5/treeherder/model/derived/jobs.py#L432)):*
Yeah, this should be in the FailureLine model as a function, rather than here.  Moved.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/1234)*
<!-- Sent from Reviewable.io -->

camd commented:
Ed: I made the changes for your initial comment.  thanks for listing that all out for me.  :)

---

Review status: 18 of 48 files reviewed at latest revision, 21 unresolved discussions.

---

*[treeherder/webapp/api/classifiedfailure.py, line 28 \[r3\]](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-K7lCCUeRfqrNpDT3f0a:-K8ROSm9hPi3Gy0zL8Uj:-1010911781) ([raw file](https://github.com/mozilla/treeherder/blob/308042f800d2d5f34865f123be04b5659457efb5/treeherder/webapp/api/classifiedfailure.py#L28)):*
removed, looks unnecessary

---

*[treeherder/webapp/api/classifiedfailure.py, line 32 \[r3\]](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-K7lCJpukAhby-E1ykbT:-K8ROVAkXEtDxU3-w3cg:1675637715) ([raw file](https://github.com/mozilla/treeherder/blob/308042f800d2d5f34865f123be04b5659457efb5/treeherder/webapp/api/classifiedfailure.py#L32)):*
removed.

---

*[treeherder/webapp/api/classifiedfailure.py, line 45 \[r3\]](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-K7lHxsrAuQSEkJn3h3u:-K8ROnvDcqENjz9_Q5qI:-894781062) ([raw file](https://github.com/mozilla/treeherder/blob/308042f800d2d5f34865f123be04b5659457efb5/treeherder/webapp/api/classifiedfailure.py#L45)):*
good catch.  :)  Fixed.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-:-K8RVP3kc-jZtdhTBuaq:1691206979)*
<!-- Sent from Reviewable.io -->

jgraham commented:
Review status: 18 of 48 files reviewed at latest revision, 22 unresolved discussions.

---

*[treeherder/model/derived/jobs.py, line 590 \[r3\]](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-K7l8wR-9pRH97uDnydi:-K8UXh7oKpMREvtMDOKz:887336412) ([raw file](https://github.com/mozilla/treeherder/blob/308042f800d2d5f34865f123be04b5659457efb5/treeherder/model/derived/jobs.py#L590)):*
I 'm not sure I agree since it returns the bug suggestions for a job, not for an artifact, and the main effect of the change is that you now need to add `with ArtifactsModel(project)` in lots of places where it wasn't previously needed. But I also don't care enough to move it back.

---

*[treeherder/model/migrations/0008__add_best_fields_to_failureline.py, line 15 \[r2\]](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-K7kzvSrNgnUxhILjl1v:-K8UIuhwRdEONPslTdIp:-1164693641) ([raw file](https://github.com/mozilla/treeherder/blob/663ca39a8b491010396479022d0793a67b26798e/treeherder/model/migrations/0008__add_best_fields_to_failureline.py#L15)):*
I don't think it's super-important, but I would be happy if we kept it.

---

*[treeherder/model/models.py, line 665 \[r2\]](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-K7ky-o22yF6-IJ3pzpp:-K8UZnN1lQZBCluy6TlM:-1479071640) ([raw file](https://github.com/mozilla/treeherder/blob/663ca39a8b491010396479022d0793a67b26798e/treeherder/model/models.py#L665)):*
Umm, these aren't the same, are they? `.get()` will throw if more than one object is returned; the replacement will silently ignore the error.

This is used to serialize the bug for the web api.

---

*[treeherder/model/models.py, line 617 \[r11\]](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-K8UXFCj-QF6L-UGODwV:-K8UXFCkd83eWZai19eQ:1865548705) ([raw file](https://github.com/mozilla/treeherder/blob/b582bea3ed266a10b97fb32310244e332711488a/treeherder/model/models.py#L617)):*
s/set/mark/ perhaps. Maybe it doesn't make too much difference though.

---

*[treeherder/webapp/api/classifiedfailure.py, line 54 \[r3\]](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-K7lJGAJsiuvAM6KvphV:-K8Uah19TcMD8iY68ePR:951120028) ([raw file](https://github.com/mozilla/treeherder/blob/308042f800d2d5f34865f123be04b5659457efb5/treeherder/webapp/api/classifiedfailure.py#L54)):*
What's the benefit of the extra complexity?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/1234)*
<!-- Sent from Reviewable.io -->

jgraham commented:
Review status: 18 of 48 files reviewed at latest revision, 22 unresolved discussions.

---

*[treeherder/webapp/api/classifiedfailure.py, line 45 \[r3\]](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-K7lHxsrAuQSEkJn3h3u:-K8ZkfJh097g7gn9Zixi:-1944366904) ([raw file](https://github.com/mozilla/treeherder/blob/308042f800d2d5f34865f123be04b5659457efb5/treeherder/webapp/api/classifiedfailure.py#L45)):*
I'm pretty surprised `.save()` doesn't check if the object is dirty.

---

*[treeherder/webapp/api/classifiedfailure.py, line 134 \[r3\]](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-K7lMl6pBafFFfVzEBIF:-K8_WKvBpdnyTHoP7l_M:-1668035802) ([raw file](https://github.com/mozilla/treeherder/blob/308042f800d2d5f34865f123be04b5659457efb5/treeherder/webapp/api/classifiedfailure.py#L134)):*
https://bugzilla.mozilla.org/show_bug.cgi?id=1241577

---

*[treeherder/webapp/api/failureline.py, line 20 \[r3\]](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-K7lNtfskrwo6l_Ux3NZ:-K8_WXcgCV0wNww7dOth:1953608263) ([raw file](https://github.com/mozilla/treeherder/blob/308042f800d2d5f34865f123be04b5659457efb5/treeherder/webapp/api/failureline.py#L20)):*
Does the base implementation use prefetch_related to ensure we do as few queries as possible? If I am reading the source correctly we need to define a queryset explicitly in order to do that.

---

*[treeherder/webapp/api/failureline.py, line 33 \[r3\]](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-K7lO9q14ieMCHQdfBSG:-K8_XQ0AJRr8NbMCj9nh:-1668035802) ([raw file](https://github.com/mozilla/treeherder/blob/308042f800d2d5f34865f123be04b5659457efb5/treeherder/webapp/api/failureline.py#L33)):*
https://bugzilla.mozilla.org/show_bug.cgi?id=1241577

---

*[treeherder/webapp/api/failureline.py, line 117 \[r3\]](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-K7lOaSW5JfG9DQg7_NQ:-K8_XTUtn817dGKovMn6:-1668035802) ([raw file](https://github.com/mozilla/treeherder/blob/308042f800d2d5f34865f123be04b5659457efb5/treeherder/webapp/api/failureline.py#L117)):*
https://bugzilla.mozilla.org/show_bug.cgi?id=1241577

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/1234)*
<!-- Sent from Reviewable.io -->

maurodoglio commented:
Review status: 18 of 48 files reviewed at latest revision, 14 unresolved discussions.

---

*[treeherder/model/derived/jobs.py, line 590 \[r3\]](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-K7l8wR-9pRH97uDnydi:-K8coGQA5Ko7ONkN_c_u:-2085570301) ([raw file](https://github.com/mozilla/treeherder/blob/308042f800d2d5f34865f123be04b5659457efb5/treeherder/model/derived/jobs.py#L590)):*
I'll leave the last call to you James.

---

*[treeherder/model/migrations/0008__add_best_fields_to_failureline.py, line 15 \[r2\]](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-K7kzvSrNgnUxhILjl1v:-K8cx5gCNm3IjlWi_oaR:1503669547) ([raw file](https://github.com/mozilla/treeherder/blob/663ca39a8b491010396479022d0793a67b26798e/treeherder/model/migrations/0008__add_best_fields_to_failureline.py#L15)):*
Whatever you guys prefer, I just wanted to bring to attention the possible data loss.

---

*[treeherder/model/models.py, line 665 \[r2\]](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-K7ky-o22yF6-IJ3pzpp:-K8ZUe80MCqH8Ko_BeHR:656466863) ([raw file](https://github.com/mozilla/treeherder/blob/663ca39a8b491010396479022d0793a67b26798e/treeherder/model/models.py#L665)):*
@camd yeah sorry, I meant Bugscache.objects.filter(id=self.bug_number).first()
@jgraham you're right, but id is a primary key so that will never happen.

---

*[treeherder/model/models.py, line 617 \[r11\]](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-K8UXFCj-QF6L-UGODwV:-K8ZU6q7ccJAihdfzkWt:-1831162067) ([raw file](https://github.com/mozilla/treeherder/blob/b582bea3ed266a10b97fb32310244e332711488a/treeherder/model/models.py#L617)):*
soft +1 to what jgraham proposes

---

*[treeherder/webapp/api/classifiedfailure.py, line 45 \[r3\]](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-K7lHxsrAuQSEkJn3h3u:-K8crR7w5wBr-OmCU9Py:733876350) ([raw file](https://github.com/mozilla/treeherder/blob/308042f800d2d5f34865f123be04b5659457efb5/treeherder/webapp/api/classifiedfailure.py#L45)):*
:crying_cat_face: 

---

*[treeherder/webapp/api/classifiedfailure.py, line 54 \[r3\]](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-K7lJGAJsiuvAM6KvphV:-K8cqcAtVUGoV5qWmCRc:458757780) ([raw file](https://github.com/mozilla/treeherder/blob/308042f800d2d5f34865f123be04b5659457efb5/treeherder/webapp/api/classifiedfailure.py#L54)):*
We discussed this on irc, we filed https://bugzilla.mozilla.org/show_bug.cgi?id=1241577 to followup on this.

---

*[treeherder/webapp/api/classifiedfailure.py, line 134 \[r3\]](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-K7lMl6pBafFFfVzEBIF:-K8csNcawicKTersk55w:-2065528475) ([raw file](https://github.com/mozilla/treeherder/blob/308042f800d2d5f34865f123be04b5659457efb5/treeherder/webapp/api/classifiedfailure.py#L134)):*
Thanks for filing that

---

*[treeherder/webapp/api/failureline.py, line 20 \[r3\]](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-K7lNtfskrwo6l_Ux3NZ:-K8ct_Dahd8MXmTH2UF-:-1880783024) ([raw file](https://github.com/mozilla/treeherder/blob/308042f800d2d5f34865f123be04b5659457efb5/treeherder/webapp/api/failureline.py#L20)):*
You're right. I'll leave to you whether to define a custom queryset or keep the current implementation

---

*[treeherder/webapp/api/failureline.py, line 30 \[r4\]](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-K8cwCgW7BzXk6jIDlc5:-K8cwCgW7BzXk6jIDlc6:-1239707223) ([raw file](https://github.com/mozilla/treeherder/blob/bea61889899ba4c68b895ee6e4a3de8700a312a8/treeherder/webapp/api/failureline.py#L30)):*
Is pk here referring to a job id?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/1234)*
<!-- Sent from Reviewable.io -->

jgraham commented:
Review status: 15 of 48 files reviewed at latest revision, 14 unresolved discussions.

---

*[treeherder/model/derived/jobs.py, line 590 \[r3\]](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-K7l8wR-9pRH97uDnydi:-K8eG5aoCltO-DiJB-1U:1311807210) ([raw file](https://github.com/mozilla/treeherder/blob/308042f800d2d5f34865f123be04b5659457efb5/treeherder/model/derived/jobs.py#L590)):*
Maybe I will move it in the future if this annoys me enough, but not now.

---

*[treeherder/model/migrations/0008__add_best_fields_to_failureline.py, line 15 \[r2\]](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-K7kzvSrNgnUxhILjl1v:-K8cyR18PLNrC2mmN-gZ:-1453726828) ([raw file](https://github.com/mozilla/treeherder/blob/663ca39a8b491010396479022d0793a67b26798e/treeherder/model/migrations/0008__add_best_fields_to_failureline.py#L15)):*
I think the decision we made was to delete all the existing FailureMatch and ClassifiedFailure rows on production and disable the tasks that would create them until this branch lands.

---

*[treeherder/webapp/api/classifiedfailure.py, line 36 \[r3\]](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-K7lGnHKXfeZG8aiMMzE:-K8dO20e6riVBGpawAfc:-1668035802) ([raw file](https://github.com/mozilla/treeherder/blob/308042f800d2d5f34865f123be04b5659457efb5/treeherder/webapp/api/classifiedfailure.py#L36)):*
https://bugzilla.mozilla.org/show_bug.cgi?id=1241577

---

*[treeherder/webapp/api/failureline.py, line 30 \[r4\]](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-K8cwCgW7BzXk6jIDlc5:-K8e6Ch1hfa60NO8-vAz:-342954011) ([raw file](https://github.com/mozilla/treeherder/blob/bea61889899ba4c68b895ee6e4a3de8700a312a8/treeherder/webapp/api/failureline.py#L30)):*
This code is now gone.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/1234)*
<!-- Sent from Reviewable.io -->

edmorley commented:
The migration on Heroku is currently failing:
https://papertrailapp.com/groups/853883/events?centered_on_id=628286458549284873&q=program%3Aapp%2Frun.4548

Presuming it just needs another `--fake` after the migration file was renamed (presuming there were no other previously unseen changes squashes into it, that hadn't yet run).

Before this merges I think we need to:
1) Merge the other migrations-only PR to master
2) Rebase this PR
3) Run the `migrate --fake` on Heroku to fix the migrate error
4) Check no exceptions/issues using New Relic
5) Try deploying this PR on stage by deploying this branch rather than merging to master (only merging to master after that shows no perf issues and no grunt build surprises)

Sound ok? :-)

---

Reviewed 17 of 23 files at r4, 1 of 1 files at r5, 3 of 6 files at r6, 1 of 1 files at r8, 2 of 6 files at r9, 2 of 3 files at r10, 1 of 1 files at r12, 2 of 2 files at r13, 2 of 2 files at r15, 1 of 1 files at r16, 1 of 1 files at r17.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[treeherder/webapp/api/failureline.py, line 111 \[r17\]](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-K8tShUTasEPD8TfDTNh:-K8tShUTasEPD8TfDTNi:-1642417004) ([raw file](https://github.com/mozilla/treeherder/blob/ae2ebb7c2442338b79f5013f998ff5db9fd4c491/treeherder/webapp/api/failureline.py#L111)):*
Why 404 -> 400?
Could you add a comment please :-)

---

*[ui/failureviewer.html, line 6 \[r17\]](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-K8tT2I75xtqm4SgPOmG:-K8tT2I75xtqm4SgPOmH:-1600446230) ([raw file](https://github.com/mozilla/treeherder/blob/ae2ebb7c2442338b79f5013f998ff5db9fd4c491/ui/failureviewer.html#L6)):*
This should be `css/failureviewer.min.css`

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/mozilla/treeherder/1234#-:-K8tR1PmBOQ7K2ShF1MW:-326777293)*
<!-- Sent from Reviewable.io -->

jgraham commented:
> Sound ok? :-)

Yes, I think that sounds reasonable.

jgraham commented:
So I think we now reached the end of stage 2. @edmorley do you want to fix up heroku or should I have a go?

edmorley commented:
> do you want to fix up heroku or should I have a go?

Fixed:
https://emorley.pastebin.mozilla.org/8857646

:-)

jgraham commented:
@camd @edmorley I've lost track of whether there are any remaining blockers for landing this.

edmorley commented:
> @camd @edmorley I've lost track of whether there are any remaining blockers for landing this.

I thought we were landing it as part of #1329?
In fact, the commits here even show up in that PR after all (hence the confusion in IRC yesterday).

That said, I'm happy regardless - my only requirement being that we try it on stage prior to merging to master and really closely look at New Relic for both exceptions and performance (I've spent so much time chasing up New Relic errors recently for things I didn't land myself, that the novelty is wearing off a but :-)).

jgraham commented:
Sure we can land both branches together if you prefer.
